test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentSelectorTest.cs

```diff
@@ -188,39 +188,48 @@ public class FullNameInAttribute
                 public string Invoke() => "Hello";
             }
         }
-
         // This will only consider types nested inside this class as ViewComponent classes
         private class FilteredViewComponentDescriptorProvider : DefaultViewComponentDescriptorProvider
         {
             public FilteredViewComponentDescriptorProvider()
-                : base(GetAssemblyProvider())
+                : this(typeof(ViewComponentContainer).GetNestedTypes(bindingAttr: BindingFlags.Public))
             {
-                AllowedTypes = typeof(ViewComponentContainer).GetNestedTypes(bindingAttr: BindingFlags.Public);
             }
 
-            public Type[] AllowedTypes { get; }
+            public FilteredViewComponentDescriptorProvider(params Type[] allowedTypes)
+                : base(GetApplicationPartManager(allowedTypes.Select(t => t.GetTypeInfo())))
+            {
+            }
 
-            protected override bool IsViewComponentType(TypeInfo typeInfo)
+            private static ApplicationPartManager GetApplicationPartManager(IEnumerable<TypeInfo> types)
             {
-                return AllowedTypes.Contains(typeInfo.AsType());
+                var manager = new ApplicationPartManager();
+                manager.ApplicationParts.Add(new TestPart(types));
+                manager.FeatureProviders.Add(new TestFeatureProvider());
+                return manager;
             }
 
-            // Need to override this since the default provider does not support private classes.
-            protected override IEnumerable<TypeInfo> GetCandidateTypes()
+            private class TestFeatureProvider : IApplicationFeatureProvider<ViewComponentFeature>
             {
-                return
-                    GetAssemblyProvider()
-                    .CandidateAssemblies
-                    .SelectMany(a => a.DefinedTypes);
+                public void PopulateFeature(IEnumerable<ApplicationPart> parts, ViewComponentFeature feature)
+                {
+                    foreach (var type in parts.OfType<IApplicationPartTypeProvider>().SelectMany(p => p.Types))
+                    {
+                        feature.ViewComponents.Add(type);
+                    }
+                }
             }
 
-            private static IAssemblyProvider GetAssemblyProvider()
+            private class TestPart : ApplicationPart, IApplicationPartTypeProvider
```

Comment:
Should we just make this a public class? It's duplicated 3-4 times that I've seen so far.